### ISSUE TITLE
Kokkos::View segfault when using runtime dimensions of 8x16 or 8x32

### ISSUE THREAD

**Issue by ldh4** [kokkos/kokkos#4906]
We found that accessing a `Kokkos::View` created with runtime dimensions of 8 produces a segfault when its dimensions are 16 or 32.

But when the view was allocated in compile time dimensions or combination of compile time and runtime dimensions, this segfault is not produced.

Here's a reproducer:
```
TEST(TEST_CATEGORY, quicktest) {
  constexpr int dim = 16;
  int idx = dim - 1;
  using ViewType = typename Kokkos::View<int********, Kokkos::DefaultExecutionSpace>;
  ViewType v(Kokkos::view_alloc("v", Kokkos::WithoutInitializing), dim, dim, dim, dim, dim, dim, dim, dim);

  Kokkos::parallel_for(1,
    KOKKOS_LAMBDA(const int&) {
      auto& lhs = v(idx, idx, idx, idx, idx, idx, idx, idx);
      lhs = 0;  // This is where it segfaults
    }
  );
}
```

* Segfault seems to happen when lhs reference is invalid as seen when running in gdb
* Usage of Kokkos::WithoutInitializing does not affect whether this segfaults or not
* Seems to segfault starting from `v(0,0,0,0,4,12,14,0)`
* This was using a serial only build and GCC 11.1.0

@nliber

**Comment by PhilMiller** [comment]
That puts the view at exactly 2^32 elements. I wonder if there's a missing conversion from int to size_t  when allocating this way. Could you report v.data(), v.size(), and &lhs

**Comment by ldh4** [comment]
When using 16 for dimensions' sizes:
`v.data(): 0x1ec6680     v.size: 4294967296      &lhs: 0x401ec667c`

For 32:
`v.data(): 0x1574680     v.size: 1099511627776   &lhs: 0x40157467c`

And this is when using 24 (a working case):
`v.data(): 0x7fdedd20e0c0        v.size: 110075314176    &lhs: 0x7fe56120e0bc`

**Comment by PhilMiller** [comment]
Fix released in 3.6.01
